text/3573-is.md

```diff
@@ -0,0 +1,284 @@
+
+An `is` operator will produce a name conflict with [the `is` method on
+`dyn Any`](https://doc.rust-lang.org/std/any/trait.Any.html#method.is) in the
+standard library, and with the (relatively few) methods named `is` in the
+ecosystem. This will not break any existing Rust code, as the operator will
+only exist in the Rust 2024 edition and newer. The Rust standard library and
+any other library that wants to avoid requiring the use of `r#is` in Rust 2024
+and newer could provide aliases of these methods under a new name; for
+instance, the standard library could additionally provide `Any::is` under a new
+name `is_type`.
+
+# Rationale and alternatives
+
+As noted in the [motivation](#motivation) section, adding the `is` operator
+allows writing pattern matches from left-to-right, which reads more naturally
+in some conditionals and fits well with method chains and similar. As noted
+under [prior art](#prior-art), other languages such as C# already have this
+exact operator for this exact purpose.
+
+We could choose not to add this operator, and have *only* `let`-chains. This
+would provide equivalent functionality, semantically; however, it would force
+pattern-matches to be written with the pattern on the left, which won't read as
+naturally in some expressions. Notably, this seems unlikely to do as effective
+a job of reducing the desire for `is_variant()` methods and helpers like
+`is_some_and(...)`.
+
+We could choose to add the `is` operator and *not* add `let`-chains. However,
+many people *already* expect `let`-chains to work as an obvious extrapolation
+from seeing `if let`/`while let` syntax.
+
+We could add this operator using punctuation instead (e.g. `~`). However, there
+is no "natural" operator that conveys "pattern match" to people (the way that
+`+` is well-known as addition). Using punctuation also seems likely to make the
+language more punctuation-heavy, less obvious, and less readable.
+
+We could add this operator using a different name. Most other names, however,
+seem likely to be longer and to fit people's expectations less well, given the
+widespread precedent of `is_xyz()` methods. The most obvious choice would be
+something like `matches`.
+
+We could permit top-level alternation in the pattern. However, this seems
+likely to produce visual and semantic ambiguity. This is technically a one-way
+door, in that `x is true|false` would parse differently depending on our
+decision here; however, the use of a pattern-match for a boolean here seems
+unlikely, redundant, and in poor style. In any case, the compiler could easily
+detect most attempts at top-level alternation and suggest adding parentheses.
+
+# Prior art
+
+`let`-chains provide prior art for having this functionality in the language.
+
+The `matches!` macro similarly provides precedent for having pattern matches in
+boolean expressions. `is` would likely be a natural replacement for most uses
+of `matches!`.
```

Comment:
I feel like this RFC should probably take a stance on `matches!`, since to me this feels like it would definitely be replaced with the new operator in a very similar way to how `try!` was replaced. This means also adding rustfix lints and exploring how generally applicable these would be.

Reply:
I've added this to the RFC, though I've kept it to the future possibilities and unresolved questions sections, because I don't want to make that a hard requirement for accepting the RFC.